tests/end-to-end/upgrade/pkg/tests/eventmesh-migration/helpers/event.go

```diff
@@ -0,0 +1,261 @@
+	labelSelector := map[string]string{
+		"function": name,
+	}
+	trigger := &v1alpha1.Trigger{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "trigger",
+			APIVersion: "eventing.knative.dev/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   name,
+			Labels: labelSelector,
+		},
+		Spec: v1alpha1.TriggerSpec{
+			Broker: "default",
+			Filter: &v1alpha1.TriggerFilter{
+				DeprecatedSourceAndType: nil,
+				Attributes:              nil,
+			},
+			Subscriber: duckv1.Destination{
+				Ref: nil,
+				URI: nil,
+			},
+		},
+	}
+
+	for _, option := range triggeroptions {
+		option(trigger)
+	}
+
+	_, err := eventingCli.Triggers(namespace).Create(trigger)
+	return err
+}
+
+func WaitForTrigger(eventingCli eventingclientv1alpha1.EventingV1alpha1Interface, subName, namespace string, retryOptions ...retry.Option) error {
+	return retry.Do(
+		func() error {
+			labelSelector := map[string]string{
+				"function": subName,
+			}
+			listOptions := metav1.ListOptions{LabelSelector: labels.SelectorFromSet(labelSelector).String()}
+			triggers, err := eventingCli.Triggers(namespace).List(listOptions)
+			if err != nil {
+				return err
+			}
+			if len(triggers.Items) == 0 {
+				return fmt.Errorf("trigger with labels: %+v  not found", labelSelector)
+			}
+
+			trigger := triggers.Items[0]
+			if !trigger.Status.IsReady() {
+				return fmt.Errorf("trigger %s not ready: %v", trigger.Name, trigger.Status)
+			}
+			return nil
+		}, retryOptions...)
+}
+
+func WaitForBroker(eventingCli eventingclientv1alpha1.EventingV1alpha1Interface, name, namespace string, retryOptions ...retry.Option) error {
+	return retry.Do(
+		func() error {
+			broker, err := eventingCli.Brokers(namespace).Get(name, metav1.GetOptions{})
+			if err != nil {
+				return err
+			}
+
+			if !broker.Status.IsReady() {
+				return fmt.Errorf("broker %s not ready: %v", name, broker.Status)
+			}
+			return nil
+		}, retryOptions...)
+}
+
+func CreateSubscription(ebCli ebClientSet.Interface, name, namespace, eventType, eventVersion, srcID string, retryOptions ...retry.Option) error {
+	subscriberEventEndpointURL := "http://" + name + "." + namespace + ".svc.cluster.local:9000/v3/events"
+	return retry.Do(func() error {
+		if _, err := ebCli.EventingV1alpha1().Subscriptions(namespace).Create(NewSubscription(name, namespace, subscriberEventEndpointURL, eventType, eventVersion, srcID)); err != nil {
+			if !strings.Contains(err.Error(), "already exists") {
+				return fmt.Errorf("error in creating subscription: %v", err)
+			}
+		}
+		return nil
+	}, retryOptions...)
+}
+
+func NewSubscription(name string, namespace string, subscriberEventEndpointURL string, eventType string, eventTypeVersion string,
+	sourceID string) *apiv1.Subscription {
+	labelsSelector := map[string]string{
+		"function": name,
+	}
+	return &apiv1.Subscription{
+		TypeMeta: metav1.TypeMeta{APIVersion: apiv1.SchemeGroupVersion.String()},
```

Comment:
Why do you define a TypeMeta in a Go object? TypeMeta is only necessary when serialization to/from JSON is involved.

Reply:
removed